docs/src/developer/topics.md

```diff
@@ -0,0 +1,387 @@
+`deepcopy` also makes a copy of an object which can be used in unsafe
+functions.
+
+Note that if `R` is a type and an element `a` of that type is passed to it,
+e.g. `R(a)` then, the Julia convention is that `a` will be returned rather
+than a copy of `a`. This convention ensures there is not an additional cost
+when coercing values that are already of the right type.
+
+We extend this convention to parent objects `R` and elements `a` of that
+parent. In particular, `R(a)` cannot be used to make a copy of `a` for use
+in an unsafe function if `R` is the parent of `a`.
+
+All other functions *may* also return the input object if they wish. In other
+words, the return value of all other functions is not suitable for use in
+an unsafe function. Only return values of arithmetic operations and `deepcopy`
+will be suitable for such use.
+
+This convention has been chosen to maximise performance of Nemo. Low level
+operations (where performance matters) make a new object, even if the result
+is the same arithmetically as one of the inputs. But higher level functions
+will not necessarily make a new object, meaning that they cannot be used with
+unsafe functions.
+
+### Aliasing rules
+
+We now summarise the aliasing rules used by Nemo and AbstractAlgebra. We are
+relatively confident by now that following these rules will result in correct
+code given the constraints mentioned above.
+
+* matrices are viewed as containers which may contain elements that alias one
+another. Other objects, e.g. polynomials, series, etc., are constructed from
+objects that do not alias one another, *even in part*.
+
+* standard unsafe operators, addeq!, mul!, addmul!, zero!, add!, are allowed
+to mutate their outputs iff the output is entirely under the control of the
+caller, i.e. it was created for the purpose of accumulation, but otherwise
+must not modify any inputs or outputs
+
+* all arithmetic functions unary minus, `+`, `-`, `*`, `^` and deepcopy must
+return new objects and cannot return one of their inputs
+
+* all other functions are allowed to return their inputs as outputs
+
+* matrix functions with an exclamation mark should not mutate the objects that
+occur as entries of the output matrix, though should be allowed to arbitrarily
+replace/swap the entries that appear in the matrix. In other words, these
+functions should be interpreted as inplace operations, rather than operations
+that are allowed to mutate the actual entries themselves.
+
+* `R(a)` where `R` is the parent of `a`, always just returns `a` and not a copy
+
+* `setcoeff!` and `setindex!` and `getcoeff` and `getindex` should not make
+copies. Note that this implies that setcoeff! should not be passed an element
+that aliases another somewhere else, even in part!
+
+* Constructors for polynomials, series and similar ring element objects (that
+are not matrices) that take an array as input, must ensure that the
+coefficients being placed into the object do not alias, even in part
```

Comment:
```suggestion
coefficients being placed into the object do not alias, even in part.
```

Reply:
Well only one of the bullet points has a full stop, so I just removed that instead.